### PR TITLE
[docs] Verify TypeScript examples in docs 

### PR DESCRIPTION
**Pull request by VanTanev**
This is an addendum to #2234. Sorry that I have the original commits in this PR, couldn't figure out a better way to do it. 

## Verify and fix all TypeScript examples in docs

This PR  also introduces `yarn verifyTypescriptExamples` command that checks every piece of TypeScript code in the docs and runs it against the TS compiler.

### CODE REVIEW

**Comment by VanTanev** [comment]
https://github.com/davidkpiano/xstate/pull/2235/commits/3d86fed43b46a2a7caf5e2011e3198736af6bbd1 and https://github.com/davidkpiano/xstate/pull/2235/commits/433887fd24ad9219c8da528fbaf685e35aa9a241 fixe the errors the script found.

Now the output is:
```
➜  xstate git:(doc/fix-typescript-guide) yarn verifyTypescriptExamples                                            
yarn run v1.22.10
$ find packages docs -iname '*.md' -not -path '*/node_modules/*' -not -path '*/.vuepress/*' -not -iname 'CHANGELOG.md' -exec scripts/verifyTypescript {} \;
docs/guides/context.md:389.ts (396,21): Cannot find name 'CounterContext'.
docs/guides/context.md:389.ts (396,37): Cannot find name 'CounterEvent'.
Done in 17.15s.
```

**Comment by VanTanev** [comment]
This is ready for review

File: docs/guides/models.md
Comment (with-heart) [inline docs/guides/models.md:102]: I know we explain later how to use the 2nd arg to `model.assign` to type the event, but that doesn't come for another 100 lines. Feels like we might be introducing a more inconvenient way to handle this first.

File: package.json
Comment (with-heart) [inline package.json:25]: Could we encapsulate this on the `bash` side instead and just run that script here? Feels like exposing implementation details in the npm script.

File: scripts/verifyTypescript.ts
Comment (with-heart) [inline scripts/verifyTypescript.ts:33]: I had that damn `non-browser context?kjj`. Always ruining my day! 😆

File: scripts/verifyTypescript.ts
Comment (with-heart) [inline scripts/verifyTypescript.ts:128]: Could be nice to add a small comment here explaining what/why. imo it's not immediately obvious.

**Review by with-heart — APPROVED** [review]
This is such a cool piece of functionality! Definitely think you should expand it into its own separate thing eventually.

File: docs/guides/models.md
Comment (VanTanev) [reply docs/guides/models.md:102]: The second argument doesn't work for named actions, though >_>

You would have to write:
```typescript
actions: {
  updateAge: userModel.assign({
    age: (_, event) => event.age
  }, 'updateAge') as any
}
```
which I believe is markedly worse.

File: package.json
Comment (VanTanev) [reply package.json:25]: The bash script takes a list of files as arguments and verifies those.

Doing it this way means you can easily run the verification for a single file, which I did for fast iteration all the time while developing:

```bash
$ scripts/verifyTypescript docs/guides/typescript.md
```

In vim, what I do is:
```
:nmap ,t :w\|!scripts/verifyTypescript docs/guides/typescript.md<cr>
```

This means, "When I press comma-T in sequence, save the current file and run the typescript verifier."

This makes for a very tight iteration loop.

File: docs/guides/models.md
Comment (with-heart) [reply docs/guides/models.md:102]: Are you sure? I use it all the time!

Comment (with-heart) [reply docs/guides/models.md:102]: Ohhhhh #2122 probably applies here. I've started setting `strictFunctionTypes: false` on all my XState projects to avoid errors in this situation for the time being, but we probably shouldn't recommend that.

File: package.json
Comment (with-heart) [reply package.json:25]: That makes sense and that's some great functionality! Love how easy it is to hook that up in vim too, gonna have to steal that :P 

I guess I'm looking at it from the perspective of:

- Would I understand what this script is supposed to do if I hadn't seen this PR and don't know `find`? The name makes that obvious but the content could be hard to parse.
- Would I know how to use this script if I hadn't seen this PR?

Might just be a detail worth noting in `CONTRIBUTING.md`!

File: docs/guides/models.md
Comment (VanTanev) [reply docs/guides/models.md:102]: It's a similar but different problem to #2122

When you use the second argument of `model.assign()`, you get a function that takes only the event with the narrowed type.  This works if used in `on: { someEvent: { actions: model.assign(..., 'someEvent') } }` - because TS expect this action definition to be for this specific event.

However, in general actions configuration, TS expects the action to be able to handle ALL possible machine events because it cannot infer which actual events call this action.

File: package.json
Comment (VanTanev) [reply package.json:25]: I would consider fixing the last 2 issues and just adding this script as a linter to all PRs, but wanted to hear some feedback from @Andarist and @davidkpiano before doing that.

File: scripts/verifyTypescript.ts
Comment (VanTanev) [reply scripts/verifyTypescript.ts:33]: Spent so much time on this before finally going "Fuck it, BAIL" :D

File: docs/guides/models.md
Comment (with-heart) [reply docs/guides/models.md:102]: Ahhhh okay, I see what you mean. `strictFunctionTypes: false` disables the check that cause either error. https://codesandbox.io/s/exciting-diffie-xbmf8?file=/src/App.tsx

File: scripts/verifyTypescript.ts
Comment (with-heart) [reply scripts/verifyTypescript.ts:33]: I'll just `jj` my way outta here!

File: scripts/verifyTypescript.ts
Comment (VanTanev) [reply scripts/verifyTypescript.ts]: Done

**Comment by VanTanev** [comment]
Should most (all?) example code be rewritten in TS?
I did that for `context.md` (not published) but then stopped because I'm not sure if this is desirable - I mean, it is for me, but not sure in general?

I'll publish the `context.md` update if everyone agrees we should move towards using TS for all example code.

**Comment by with-heart** [comment]
> Should most (all?) example code be rewritten in TS?

@VanTanev I think so because:

- types provide context for the examples even if you don't end up using them yourself
- it's significantly easier to read ts and understand which parts to remove if using js than it is to read js and mentally keep track of what all of the things are supposed to be

It'd also be fairly easy to hook up a `remark` plugin like [`remark-typescript`](https://github.com/trevorblades/remark-typescript) to add the ability to switch between them if we wanted or something.

**Comment by VanTanev** [comment]
OK, added the `condext.md` TS rewrite!

**Comment by VanTanev** [comment]
I also added the ability to verify JS, but have not fixed the errors that show up:
```
➜  xstate git:(doc/fix-typescript-guide) ENABLE_JS=1 yarn verifyCodeExamples                                                                                                                                                                            
yarn run v1.22.10
$ find packages docs -iname '*.md' -not -path '*/node_modules/*' -not -path '*/.vuepress/*' -not -iname 'CHANGELOG.md' -exec scripts/verifyCodeExamples "{}" + ;
packages/xstate-inspect/README.md:148.ts (152,1): Cannot find name 'windowReceiver'.
packages/xstate-inspect/README.md:148.ts (156,1): Expression expected.
packages/xstate-inspect/README.md:176.ts (178,5): Cannot find name 'inspect'.
packages/xstate-inspect/README.md:47.ts (49,1): Cannot find name 'inspect'.
packages/xstate-inspect/README.md:47.ts (55,1): Cannot find name 'inspect'.
packages/xstate-inspect/README.md:66.ts (68,3): Cannot find name 'inspect'.
packages/xstate-inspect/README.md:73.ts (75,3): Cannot find name 'inspect'.
packages/xstate-svelte/README.md:244.ts (248,3): Argument of type 'string | null' is not assignable to parameter of type 'string'.
  Type 'null' is not assignable to type 'string'.
packages/xstate-svelte/README.md:244.ts (251,25): Cannot find name 'useMachine'.
packages/xstate-svelte/README.md:244.ts (251,36): Cannot find name 'someMachine'.
packages/xstate-immer/README.md:126.ts (146,13): Object is of type 'unknown'.
packages/xstate-immer/README.md:126.ts (146,45): Property 'value' does not exist on type 'EventObject'.
packages/xstate-immer/README.md:187.ts (197,3): Object is of type 'unknown'.
packages/xstate-immer/README.md:187.ts (201,3): Object is of type 'unknown'.
packages/xstate-immer/README.md:187.ts (222,17): Cannot find name 'interpret'.
packages/xstate-immer/README.md:44.ts (49,3): Object is of type 'unknown'.
packages/xstate-immer/README.md:44.ts (66,36): Object is of type 'unknown'.
packages/xstate-fsm/README.md:194.ts (195,19): Cannot find name 'machine'.
packages/xstate-fsm/README.md:226.ts (227,21): Cannot find name 'machine'.
packages/xstate-fsm/README.md:226.ts (228,18): Cannot find name 'machine'.
packages/xstate-fsm/README.md:226.ts (229,20): Cannot find name 'machine'.
packages/xstate-fsm/README.md:255.ts (258,31): Argument of type '{}' is not assignable to parameter of type 'Config<object, EventObject, { value: any; context: object; }>'.
  Type '{}' is missing the following properties from type 'Config<object, EventObject, { value: any; context: object; }>': initial, states
packages/xstate-fsm/README.md:379.ts (401,47): Property 'redLights' does not exist on type 'object'.
packages/xstate-fsm/README.md:84.ts (87,37): Argument of type '{}' is not assignable to parameter of type 'Config<object, EventObject, { value: any; context: object; }>'.
  Type '{}' is missing the following properties from type 'Config<object, EventObject, { value: any; context: object; }>': initial, states
packages/xstate-graph/README.md:15.ts (19,17): Expected 1-2 arguments, but got 0.
packages/xstate-graph/README.md:172.ts (174,23): ';' expected.
packages/xstate-graph/README.md:172.ts (175,12): ';' expected.
packages/xstate-graph/README.md:172.ts (175,14): Cannot find name 'State'.
packages/xstate-graph/README.md:172.ts (175,20): ';' expected.
packages/xstate-graph/README.md:172.ts (175,22): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (175,27): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (176,12): ';' expected.
packages/xstate-graph/README.md:172.ts (179,20): Cannot find name 'State'.
packages/xstate-graph/README.md:172.ts (179,26): ',' expected.
packages/xstate-graph/README.md:172.ts (179,32): Expression expected.
packages/xstate-graph/README.md:172.ts (180,18): ',' expected.
packages/xstate-graph/README.md:172.ts (181,9): ',' expected.
packages/xstate-graph/README.md:172.ts (181,10): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (183,18): ';' expected.
packages/xstate-graph/README.md:172.ts (183,20): Cannot find name 'State'.
packages/xstate-graph/README.md:172.ts (183,26): ';' expected.
packages/xstate-graph/README.md:172.ts (183,28): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (183,33): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (184,18): ';' expected.
packages/xstate-graph/README.md:172.ts (184,28): ';' expected.
packages/xstate-graph/README.md:172.ts (184,30): Left side of comma operator is unused and has no side effects.
packages/xstate-graph/README.md:172.ts (184,54): ';' expected.
packages/xstate-graph/README.md:172.ts (185,10): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (186,9): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (187,7): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (187,8): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (188,7): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (189,5): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (190,4): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (191,3): Declaration or statement expected.
packages/xstate-graph/README.md:172.ts (192,1): Declaration or statement expected.
packages/xstate-graph/README.md:301.ts (304,17): Cannot find name 'createMachine'.
packages/xstate-graph/README.md:301.ts (310,7): Left side of comma operator is unused and has no side effects.
packages/xstate-graph/README.md:301.ts (311,3): Cannot find name 'stateNode'.
packages/xstate-graph/README.md:301.ts (311,12): ';' expected.
packages/xstate-graph/README.md:301.ts (311,29): Declaration or statement expected.
packages/xstate-graph/README.md:301.ts (312,13): Left side of comma operator is unused and has no side effects.
packages/xstate-graph/README.md:301.ts (317,3): Cannot find name 'edges'.
packages/xstate-graph/README.md:301.ts (317,8): ';' expected.
packages/xstate-graph/README.md:301.ts (318,24): Expression expected.
packages/xstate-graph/README.md:301.ts (318,43): Expression expected.
packages/xstate-graph/README.md:301.ts (318,67): Expression expected.
packages/xstate-graph/README.md:335.ts (344,48): Object is of type 'unknown'.
packages/xstate-graph/README.md:335.ts (344,62): Property 'value' does not exist on type 'EventObject'.
packages/xstate-graph/README.md:335.ts (351,23): Cannot find name 'getShortestPaths'.
packages/xstate-graph/README.md:50.ts (52,23): ';' expected.
packages/xstate-graph/README.md:50.ts (53,12): ';' expected.
packages/xstate-graph/README.md:50.ts (53,14): Cannot find name 'State'.
packages/xstate-graph/README.md:50.ts (53,20): ';' expected.
packages/xstate-graph/README.md:50.ts (53,22): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (53,27): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (54,11): ';' expected.
packages/xstate-graph/README.md:50.ts (56,18): Cannot find name 'State'.
packages/xstate-graph/README.md:50.ts (56,24): ',' expected.
packages/xstate-graph/README.md:50.ts (56,30): Expression expected.
packages/xstate-graph/README.md:50.ts (57,16): ',' expected.
packages/xstate-graph/README.md:50.ts (58,7): ',' expected.
packages/xstate-graph/README.md:50.ts (58,8): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (60,16): ';' expected.
packages/xstate-graph/README.md:50.ts (60,18): Cannot find name 'State'.
packages/xstate-graph/README.md:50.ts (60,24): ';' expected.
packages/xstate-graph/README.md:50.ts (60,26): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (60,31): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (61,16): ';' expected.
packages/xstate-graph/README.md:50.ts (61,26): ';' expected.
packages/xstate-graph/README.md:50.ts (61,28): Left side of comma operator is unused and has no side effects.
packages/xstate-graph/README.md:50.ts (61,52): ';' expected.
packages/xstate-graph/README.md:50.ts (62,8): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (63,7): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (64,5): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (65,4): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (66,3): Declaration or statement expected.
packages/xstate-graph/README.md:50.ts (67,1): Declaration or statement expected.
Cannot resolve @xstate/vue
/home/ivan/dev/xstate/node_modules/esbuild-register/dist/node.js:2699
        throw error;
        ^

Error: Debug Failure. False expression.
    at processImportedModules (/home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:108359:26)
    at findSourceFileWorker (/home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:108160:17)
    at findSourceFile (/home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:108021:26)
    at /home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:107973:85
    at getSourceFileFromReferenceWorker (/home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:107940:34)
    at processSourceFile (/home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:107973:13)
    at processRootFile (/home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:107783:13)
    at /home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:106665:67
    at Object.forEach (/home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:384:30)
    at Object.createProgram (/home/ivan/dev/xstate/node_modules/typescript/lib/typescript.js:106665:16)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: docs/guides/context.md
Comment (davidkpiano) [inline docs/guides/context.md:23]:
```suggestion
const glassIsFull = (context: Context) =>
```

**Review by davidkpiano — APPROVED** [review]
This is really cool (sorry, finally getting to this). Apart from 1 suggested change, LGTM, let me know what you think!

File: docs/guides/context.md
Comment (VanTanev) [reply docs/guides/context.md:23]: I thought it would be nice to show off the `ConditionPredicate` type, but can do the simple version too.

Comment (davidkpiano) [reply docs/guides/context.md:23]: > I thought it would be nice to show off the `ConditionPredicate` type, but can do the simple version too.

Let's do the simple version, as I think `ConditionPredicate` seems like extra things you need to know (mental overload) and the name of that type might change in v5 anyway (to `GuardPredicate` or something like that).

File: package.json
Comment (Andarist) [reply package.json]: Even though this is in bash which I barely know it's still somewhat readable. I wouldn't be able to tweak this beyond some simple copy-pasta-like changes - but I guess there might just not be a need for this any time soon.

TLDR: I'm OK with this for now.

**Comment by Andarist** [comment]
q: would https://github.com/shikijs/twoslash provide similar functionality (and more)? I don't intend to block this PR - it is useful on its own and we don't want to pursue integrating `twoslash` right away. Just thought that maybe you have already looked into it - because I didn't have much time to do that myself - and that maybe you would already have some opinions on it that could be shared.

**Comment by VanTanev** [comment]
> q: would https://github.com/shikijs/twoslash provide similar functionality (and more)?

It does!

However, it wasn't released at the time I wrote this PR, and integration seems non-trivial.
Do you think the PR is worth merging as is, or should I try to instead integrate `twoslash`?

**Comment by VanTanev** [comment]
This is ready for re-review. I think it would be useful to merge as is and consider moving to `twoslash` in the future when someone has the time to do that.

File: examples/fetch/package.json
Comment (davidkpiano) [inline examples/fetch/package.json:10]: Can we remove this?

**Review by davidkpiano — COMMENTED** [review]
My only question now is how approachable these examples are going to be to those using JavaScript, and not TypeScript? Should we find a way to allow developers to toggle between the two?

**Comment by VanTanev** [comment]
> My only question now is how approachable these examples are going to be to those using JavaScript, and not TypeScript? Should we find a way to allow developers to toggle between the two?

Let's leave that for a separate PR. There are fixes in here that will be useful to xstate users today. We can strive to make it even better tomorrow.

File: examples/fetch/package.json
Comment (VanTanev) [reply examples/fetch/package.json:10]: It is required for the root clean script to run, otherwise it fails. In turn, the typescript verification script requires the root clean script.

Comment (VanTanev) [reply examples/fetch/package.json:10]: Actually, scratch that, the verify script no longer requires this (it did at some point). 

However, if you're OK I think we should still leave this in, because it does fix the root clean script.

**Comment by davidkpiano** [comment]
> > My only question now is how approachable these examples are going to be to those using JavaScript, and not TypeScript? Should we find a way to allow developers to toggle between the two?
> 
> Let's leave that for a separate PR. There are fixes in here that will be useful to xstate users today. We can strive to make it even better tomorrow.

It will help some users but will confuse others, see here: https://twitter.com/DavidKPiano/status/1404806818832977927

I want to be cautious about merging this unless we can satisfy _both_ JS and TS users, and I think the best way to do it would be to show JS and TS together, via something like this: https://github.com/phryneas/remark-typescript-tools

We cannot merge this to `main` as-is unless we have both languages.

**Comment by VanTanev** [comment]
OK. I don't have the bandwidth to do both right now, not sure when I will.

**Comment by davidkpiano** [comment]
> OK. I don't have the bandwidth to do both right now, not sure when I will.

We can take this as a starting point and work with it; thanks so much for the work on this so far!

**Pull request closed without merging**
